File: coding_skill_cpp/26_hal_pattern.cpp
/* C++ HAL - 추상 클래스 */
#include <iostream>
using namespace std;

class IGPIO {
public:
    virtual ~IGPIO() = default;
    virtual void write(int pin, bool value) = 0;
};

class STM32_GPIO : public IGPIO {
public:
    void write(int pin, bool value) override {
        cout << "[STM32] GPIO " << pin << " = " << value << endl;
    }
};

class Nordic_GPIO : public IGPIO {
public:
    void write(int pin, bool value) override {
        cout << "[Nordic] GPIO " << pin << " = " << value << endl;
    }
};

void app_toggle_led(IGPIO& hal) {
    hal.write(5, true);
}

int main() {
    cout << "=== C++ HAL ===" << endl;
    STM32_GPIO stm32;
    Nordic_GPIO nordic;
    
    app_toggle_led(stm32);
    app_toggle_led(nordic);
    
    return 0;
}

File: coding_skill_py/26_hal_pattern.py
"""Python HAL - ABC"""
from abc import ABC, abstractmethod

class IGPIO(ABC):
    @abstractmethod
    def write(self, pin: int, value: bool):
        pass

class STM32_GPIO(IGPIO):
    def write(self, pin: int, value: bool):
        print(f"[STM32] GPIO {pin} = {value}")

class Nordic_GPIO(IGPIO):
    def write(self, pin: int, value: bool):
        print(f"[Nordic] GPIO {pin} = {value}")

def app_toggle_led(hal: IGPIO):
    hal.write(5, True)

def main():
    print("=== Python HAL ===")
    stm32 = STM32_GPIO()
    nordic = Nordic_GPIO()
    
    app_toggle_led(stm32)
    app_toggle_led(nordic)

if __name__ == "__main__":
    main()
